[quant][eagermode] Custom module support

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44835 [quant][eagermode] Custom module support**

Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

jerryzh168 committed:
[quant][eagermode] Custom module support

Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -429,7 +451,10 @@ def _convert(module, mapping=None, inplace=False):
-        if type(mod) not in SWAPPABLE_MODULES:
+        # both swappable modules and observed custom modules are
+        # swapped as one unit
+        if type(mod) not in SWAPPABLE_MODULES and \
```
When we support custom module fusion, we will need to change code here. Currently, it looks like Swappable modules defines the list of fusions that are supported. Thoughts on how this can be addressed? There main use case would be : nn.module A + nn.module B => Custom Fused Module C

jerryzh168 replied:
we can make this configuration with registrations as well, can be done in a separate PR before we need this feature

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -571,6 +575,115 @@ def forward(self, indices, offsets, per_sample_weights, linear_in):
         self.checkDynamicQuantizedModule(quantized_model.emb, torch.nn.quantized.EmbeddingBag, torch.quint8)
 
+    @skipIfNoFBGEMM
+    def test_custom_module_class(self):
```
Similar request as in the previous PR, lets add tests for each of the workflows (PTQ, Dynamic and QAT) and also for modules with multiple input/outputs (LSTMs).

jerryzh168 replied:
for dynamic/qat, will do that after the mixed mode support change.
lstm test will be done after zafar write the lstm pr. but I'll add multiple input/output support in a separate PR

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

z-a-f commented on torch/quantization/quantize.py:
```diff
@@ -429,7 +451,10 @@ def _convert(module, mapping=None, inplace=False):
-        if type(mod) not in SWAPPABLE_MODULES:
+        # both swappable modules and observed custom modules are
+        # swapped as one unit
+        if type(mod) not in SWAPPABLE_MODULES and \
```
nit: will the `SWAPPABLE_MODULES` be a globa variable, or are we planning on replacing it with a getter function?

jerryzh168 replied:
yeah we can do that when there is a need

z-a-f reviewed (approved):
LGTM

raghuramank100 reviewed (approved):
If we can add more tests like requested in the other PR for both eager and IR custom module support, that would be great.

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][eagermode] Custom module support"


Summary:
This is for feature parity with fx graph mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23745086](https://our.internmc.facebook.com/intern/diff/D23745086)

[ghstack-poisoned]
